File: backend/dotnet/Controllers/AuthController.cs
using dotnet.Interfaces;
using Microsoft.AspNetCore.Mvc;

namespace dotnet.Controllers
{
    [ApiController]
    [Route("auth")]
    public class AuthController : ControllerBase
    {
        private readonly IAuthService _authService;

        public AuthController(IAuthService authService)
        {
            _authService = authService;
        }

        // Logging in

        [HttpPost("login")]
        public IActionResult Login([FromBody] AuthRequestData request)
        {
            try
            {
                string token = _authService.Login(request.email, request.password);
                return Ok(new AuthResponseData { token = token });
            }
            catch (Exception ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
        }

        [HttpPost("logout")]
        public IActionResult Logout([FromBody] LogoutRequestData payload)
        {
            string SessionId = payload.verificationToken;

            _authService.Logout(SessionId);
            return Ok("Logged out successfully");
        }

        // Registering

        [HttpPost("register")]
        public IActionResult Register([FromBody] AuthRequestData payload)
        {
            try
            {
                _authService.Register(payload.email, payload.password);
                return Ok("Email Sent Successfully");
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                return BadRequest(e.Message);
            }
        }

        [HttpPost("register/verification")]
        public IActionResult RegisterVerification([FromBody] LogoutRequestData payload)
        {
            try
            {
                string VerificationToken = payload.verificationToken;

                _authService.RegisterVerification(VerificationToken);
                return Ok("Verification successful");
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                return BadRequest(e.Message);
            }
        }

        // Password Reset

        [HttpPost("password-reset")]
        public IActionResult PasswordReset([FromBody] PasswordResetRequestData payload)
        {
            try
            {
                _authService.PasswordReset(payload.email);
                return Ok("Password reset link sent to email");
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                return BadRequest(e.Message);
            }
        }

        [HttpPost("password-reset/verification")]
        public IActionResult PasswordResetVerification([FromBody] PasswordResetVerificationRequestData payload)
        {
            try
            {
                _authService.PasswordResetVerification(payload.resetPasswordToken);
                return Ok("Valid token");
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                return BadRequest(e.Message);
            }
        }

        [HttpPost("password-reset/confirmation")]
        public IActionResult PasswordResetConfirmation([FromBody] PasswordResetConfirmationRequestData payload)
        {
            try
            {
                _authService.PasswordResetConfirmation(payload.resetPasswordToken, payload.newPassword);
                return Ok("Password reset successful");
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                return BadRequest(e.Message);
            }
        }
    }

    // DTOs
    public class AuthRequestData
    {
        public string email { get; set; }
        public string password { get; set; }
    }

    public class AuthResponseData
    {
        public string token { get; set; }
    }

    public class LogoutRequestData
    {
        public string verificationToken { get; set; }
    }

    public class PasswordResetRequestData
    {
        public string email { get; set; }
    }

    public class PasswordResetVerificationRequestData
    {
        public string resetPasswordToken { get; set; }
    }

    public class PasswordResetConfirmationRequestData
    {
        public string resetPasswordToken { get; set; }
        public string newPassword { get; set; }
    }
}

File: backend/springboot/src/main/java/tn/maiko26/springboot/controllers/AuthController.java
package tn.maiko26.springboot.controllers;

import lombok.Data;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import tn.maiko26.springboot.services.AuthService;

import java.util.Map;

@RestController
@RequestMapping("/auth")
public class AuthController {

    @Autowired
    private AuthService authService;

    // Logging in

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody AuthRequestData request) {
        try {
            String token = authService.login(request.getEmail(), request.getPassword());
            return ResponseEntity.ok().body(new AuthResponse(token));
        } catch (Exception e) {
            return ResponseEntity.status(401).body(e.getMessage());
        }
    }

    @PostMapping("/logout")
    public ResponseEntity<?> logout(@RequestBody Map<String, String> payload) {
        String sessionId = payload.get("verificationToken");

        authService.logout(sessionId);
        return ResponseEntity.ok().body("Logged out successfully");
    }

    // Registering

    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody AuthRequestData request) {
        try {
            authService.register(request.getEmail(), request.getPassword(),request.getName());
            return ResponseEntity.ok().body(Map.of("message", "Email Sent Successfully"));
        } catch (Exception e) {
            e.printStackTrace();
            return ResponseEntity.status(400).body(Map.of("message", e.getMessage()));
        }
    }

    @PostMapping("/register/verification")
    public ResponseEntity<?> registerVerification(@RequestBody Map<String, String> payload) {
        try {
            String verificationToken = payload.get("verificationToken");

            authService.registerVerification(verificationToken);
            return ResponseEntity.ok().body(Map.of("message", "Verification successful"));
        } catch (Exception e) {
            return ResponseEntity.status(400).body(Map.of("message", e.getMessage()));
        }
    }

    // Password Reset

    @PostMapping("/password-reset")
    public ResponseEntity<?> passwordReset(@RequestBody Map<String, String> payload) {
        try {
            String email = payload.get("email");

            authService.passwordReset(email);
            return ResponseEntity.ok().body(Map.of("message", "Password reset link sent to email"));
        } catch (Exception e) {
            return ResponseEntity.status(400).body(Map.of("message", e.getMessage()));
        }

    }

    @PostMapping("/password-reset/verification")
    public ResponseEntity<?> passwordResetVerification(@RequestBody Map<String, String> payload) {
        try {
            String resetPasswordToken = payload.get("resetPasswordToken");

            authService.passwordResetVerification(resetPasswordToken);
            return ResponseEntity.ok().body(Map.of("message", "Valid token"));
        } catch (Exception e) {
            return ResponseEntity.status(400).body(Map.of("message", e.getMessage()));
        }

    }

    @PostMapping("/password-reset/confirmation")
    public ResponseEntity<?> passwordResetConfirmation(@RequestBody PasswordResetRequestDTO request) {
        try {
            authService.passwordResetConfirmation(request.getResetPasswordToken(), request.getNewPassword());
            return ResponseEntity.ok().body(Map.of("message", "Password reset successful"));
        } catch (Exception e) {
            return ResponseEntity.status(400).body(Map.of("message", e.getMessage()));
        }

    }


    @ExceptionHandler
    public ResponseEntity<?> conflict() {
        return ResponseEntity.status(500).body("Internal Server Error");
    }
}

@Data
class AuthRequestData {
    private String email;
    private String password;
    private String name;
}

@Data
class PasswordResetRequestDTO {
    private String newPassword;
    private String resetPasswordToken;
}

class AuthResponse {
    private String token;

    public AuthResponse(String token) {
        this.token = token;
    }

    public String getToken() {
        return token;
    }
}
